Bootsy has no inherent dependency on Rails, conditionally require it

Benjamin-Dobell opened the pull request:
Bootsy is predominantly a Rails extension (views etc.). However, the data model components (ActiveModel & CarrierWave) of Bootsy are not specifically tied to Rails and can be used without it.

A client of mine, for which I just integrated Bootsy, has their data model in a Gem, shared across various projects; only one of which is Rails, the rest are Sinatra etc. (ActiveModel & CarrierWave, but nothing else Rails).

Consequently, this pull request conditionally requires the Bootsy Rails Engine.

volmer commented:
Bootsy was designed as a Rails engine, and IMO it only makes sense as such. Allowing it to be usable elsewhere would require a lot more than that (documentation, tests, proper view/asset handling), things that are beyond the scope of this project. In your case, what about installing it with `require: false` and requiring only the classes/modules you need?

Benjamin-Dobell commented:
Unfortunately this approach doesn't work.

Projects only contain explicit dependencies in their Gemfile. It's the "models" Gem that depends on Bootsy via a Gemspec, each individual project doesn't depend on Bootsy directly. There is no mechanism to disable auto require from a Gemspec, at least that I'm aware of.

To be clear, I'm not suggesting that Bootsy makes sense without a Rails project at all. It is the client's Rails project that utilises Bootsy's views. I'm just after a way to access (read) Bootsy generated content from non-Rails projects.

This is important because micro services are generally deemed to be superior to monolithic projects for various reasons (performance, maintainance etc.) In order to build micro services the data model typically needs to be shared. The data model requires knowledge of Bootsy, in particular `ImageUploader` to work.

For instance, the client has a Sinatra web services project which communicates with mobile apps, other micro services and web browser AJAX; from third-party websites embedding the client's JavaScriot API in their own (likely non-rails) websites. The Web services needs to be able to access the database, including the HTML content (created by Bootsy) and serve this data to clients.

If Bootsy's models, in my case, only `ImageUploader`, were separated out into its own Gem then that could be depended on by both Bootsy and other projects. However, Bootsy itself is quite lightweight so it seemed like overkill, this pull request was simply the path of least resistance.

Benjamin-Dobell commented:
Hmm, in hindsight I think I was incorrect about Gemspecs and autorequire. Given autorequire is a Bundler thing then Gems included in a Gemspec shouldn't be required automatically at all.

I'll take a go at requiring just what I need. Although it does make.me uncomfortable as it's prone to break.

Benjamin-Dobell commented:
Okay, so *now* I remember why that approach wouldn't work.

I can't just go requiring `ImageLoader` (even after #220 is merged), because ImageLoader has a dependency on the `Bootsy` module, `Bootsy.storage` etc. However, the `Bootsy` module is defined in `bootsy.rb`, which I can't require as it requires `bootsy/engine.rb`.

Benjamin-Dobell commented:
Okay, yeah this was proving to be too difficult with the way Bootsy is currently designed.

I've removed my dependency on Bootsy from my models Gem and added it in the Rails project only. I reopen my model(s) in the Rails project and add `include Bootsy::Container` there. This probably wouldn't work so well with ActiveRecord, but seems as this project is using Mongoid I can get away with it.

The ideal solution would be to split all the ActiveModel stuff in Bootsy out into its own Gem, without any ActionController or ActionView stuff. That way Bootsy could depend on that Gem, and then non-Rails consumers could just depend on that Gem rather than Bootsy itself. It's hard for me to submit a pull request like that, and you may simply not want to, so I'll leave it up to you.
